### PR TITLE
cherry picked changes for sdram edge support

### PR DESCRIPTION
**Pull request by alan-stokes**
cherry picked changes to support SDRAM edge allocation from roberts ear modifications

depends upon:
https://github.com/SpiNNakerManchester/SpiNNFrontEndCommon/pull/492

### CODE REVIEW

**Comment by alan-stokes** [comment]
so this is brought from the dead version of this pr
@rowleya 
The OutgoingEdgePartition for multicast is designed to make multicast routing easier (since this is pretty much a multicast partition of the edges), but it isn't clear that the SDRAM OutgoingEdgePartition makes any sense. It might be better if this was a "partition" (or group?) of edges that all share some part of SDRAM, as then the allocation would actually be quite easy.

a outgoing edge partition is fundamentally a group of edges. Making a seperate thing for sdram when the partition has a traffic type already seems like making a api for no good reason. 

the group of sdram edges can be 1 of 2 types. consistent block or segmented block. That means there's actually context changes based on that for what the address is for each edge. so just a group makes no sense either, as it needs to do logic. In the use cases i've seen the segregated version also assumes its 1 full continuous block of sdram, as they can read it all (nengo ensambles) and so the alloation is on the group, not the edge. 

basically, i think this way is easier and more clear and we already had said api......

**Comment by Christian-B** [comment]
I think we need to come up with a toy example that demonstrates / tests the ideas here outside of the whole ear.

Without comments on the classes and interfaces it is hard to see exactly what they are intended for,

**Comment by alan-stokes** [comment]
have a look at lucas branch. hes also using sdram edges. or nengo. 

but adding a test case for this specifically im all for

**Comment by Christian-B** [comment]
The more I think about it the SDRAM Edge is not a real object.  The work of writing to SDRAM and reading from SDRAM is done by vertexes.
These vertexes can ONLY work in the SINGLE specific way connected to other vertexes that only work in the SAME SINGLE specific way.

So what we really have is two vertexes with a on same chip requirement and NO edge between then.

**Comment by rowleya** [comment]
So a possible better model then is a "SharedSDRAM" object of some sort.  This might be thought of as like an edge, or not depending on your viewpoint.  It would however have the property of linking together several vertices, which then of course have a requirement of being on the same chip.  So the model is an object which has a size and a list of vertices that share that space.

**Comment by Christian-B** [comment]
I disagree with @rowleya on having an None Edge "SharedSDRAM" object of some sort.

In my opinion the best is No object at all because it is tied to close to the Vertex.  But if you have an Object between two Vertexes than it is an Edge!

**Comment by alan-stokes** [comment]
"For SDRAM edges to be useful we need an example where the same Vertexes can have different types or at least behavior of edge without changing the vertex or the parameters passed into the vertex."

that's true, there are 2 different behaviours. and if you go down the route of putting on the vertex's head. who pre chance. allocates the sdram to its resources? the src? the destination? how you comm between them. oh yea, that's EXACTLY what a communication edge looks like, and why we have seperate resource operations on edges. such as mc key allocation. WE DON'T MAKE VERTICES DO THAT! So its a edge who's communication type is SDRAM. It is not vertex specific at that point. 

I holy disagree with @Christian-B viewpoint here. 

nengo uses the sdram edge in a seqmented way but where it needed it to be a continous block. robert uses it as a ring buffer link process (where theres a set of buffers in the sdram, and those buffers are next to each other for ease. It could be segmented, or consisdered a single sdram edge with x * n buffers worth of sdram and allocate internally. which is what i deicded to do that time.

**Comment by alan-stokes** [comment]
heck. if you make it vertex specific your going to have to make vertices ask the spinnman for its allocation at some point? or is it a dsg region which has no input data? how the heck do you comm to the otehr vertex the sdram address? given it cant access the dsg map of the otehr core on boot easily.

Your view puts you in a horrbile chicken vs egg issue. whereas having the sdram allocated seperately to everything else, and having it considered a graph resoruce makes your life a TON easier.

**Comment by rowleya** [comment]
I think the issue comes down to whether this shared SDRAM will be initially filled with data, or just a space that will be used in some way by 2 or more vertices.  

I think Christian's point is that it isn't that hard to just have one of the vertices specify and, if necessary, fill the space using the parameters given to it, rather than have the space on an edge and then the edge has to get the parameters from one or other of the vertices, meaning more indirection.

I think Alan's point is that given a shared space, it is difficult to communicate the location of the space to both vertices.  His idea is then to have something allocate the space in advance and then each vertex can be given the address at DSG time.  This is fine if the space is to be empty but could be more complicated if the space is to be filled using parameters from the vertex in advance.  There is a question of whether we have a use-case where the space is filled; I believe all the cases to date have the space empty at the start and then fill it during execution.

I note that SCAMP/SARK does provide a mechanism through which allocated SDRAM spaces can be shared using SDRAM tags (not to be confused with IP Tags).  I believe that this is how most users to date have been solving this issue.  The next issue with this is that the vertices have to agree in advance which tags they are going to use.  So far this has been simple enough since there haven't been any cross-overs between things that use the SDRAM tags, but I can foresee a case where this will no longer be safe: that of the combination of the SpiNNakEar model with the multi-core neuron processing model, especially if placement puts components of each on the same chip.  An additional case would be having two sets of vertices for multi-core neurons on the same chip.  In these cases it would be useful to at least assign the SDRAM tags to be used in advance so that the vertices can then communicate effectively.

So what I propose is something that allows vertices to say they need a resource which shares a space in SDRAM, and are therefore assigned an SDRAM tag to make use of the space.  This *doesn't* need the space to be allocated in advance though, so the size of the space *could* still be included in the resources of one of the vertices.  That said, it might be as easy to specify this size as part of the SDRAM tag requirement; this I am not sure about, but am happy to hear opinions.

**Comment by alan-stokes** [comment]
So to build on your views there rowley. 
1. 
"I think Christian's point is that it isn't that hard to just have one of the vertices specify and, if necessary, fill the space using the parameters given to it, rather than have the space on an edge and then the edge has to get the parameters from one or other of the vertices, meaning more indirection."

that would be more suitable to be described as a shared DSG region. not a sdram region. It has the same issue of needing some way to update the app pointer tables on the given vertices. But that would be a much better representation of that view. 

2. 
"I think Alan's point is that given a shared space, it is difficult to communicate the location of the space to both vertices. His idea is then to have something allocate the space in advance and then each vertex can be given the address at DSG time. This is fine if the space is to be empty but could be more complicated if the space is to be filled using parameters from the vertex in advance. There is a question of whether we have a use-case where the space is filled; I believe all the cases to date have the space empty at the start and then fill it during execution."

this is exactly what every use case ive seen so far does. how that space is then used can be different. But i have never seen any usecase where its a shared parameter space. 

3. There are 3 reasons i dont like using the sdram tags. 
1. its the limited number of them which will result in problems down line. Therefore avoiding them as much as possible is better.
2. Someone needs to allocate that sdram. which means theres a master and slave process. Which is clunky at best, coz now the end user needs to decide who of the potentially many cores in the communication needs to be the lead and allocate it. Which is also true in the python, who allocates it? if the size is based off his destinations (such as a segmented version, (remind you of nengo) then you cant know at resource allocation time what the size is. coz you cant ensure your children have been partitioned yet). Thus the chicken and egg issue. 
3. this would mean that the same chip as constraint needs to be added manually. whereas a sdram edge which infers the same chip as means less wrok for the end users.

The reason I really dont want the vertex toi have to allcate the sdram into one of its resoruces is 2 fold. 
1. adds that master slave fiasco
2. the vertices to date, only operate on 2 aspects.
1. for partitioning, which uses space for init DATA.
2. for formatting that init DATA. 

no vertex to date, has had to do allocation for a communication component. we dont ask them to allocate basekeys and the range of those keys. we ask for n keys. or make special cases for things that NEED SPECIFIC keys, which is the mc equivalent. The tools allocate keys and give them the id. My stuff keeps to that ethos by allocating the sdram away from the vertex, and allowing the vertex to get the minimum stuff it needs to use it. aka the size and base address. No vertex will need a SPECIFIC sdram address, and if they do. I want to understand that use case majorly. 

So putting it in a SDRAM edge makes the most clear sense. You discussed bi directional stuff. yes, id currently give you 2 sets of sdram. 1 for each direction. I guess we could someone fudge it to be one, if the requirement occurred. But i remind you that if you had a bidirectional MC edge, it would have 2 keys. The entire graph setup assumes unidirectional comms. and so until such a use case of sdram bi-directional NEEDING the same space and needing to be in the graph as both direction comes to pass. im inclined to ignore it.

**Comment by rowleya** [comment]
OK, so *my* main objections here are still the use of OutgoingEdgePartition with SDRAM.  My reasons are:
 1. I don't think this acts like multicast, so it doesn't make sense to have single-source-multiple-target in the definition
 1. I *do* have an example of multiple-source-single-target (which is what the multi-core neuron model does I am pretty sure).
 1. Storing the SDRAM Partition along with the multicast partition means that you have to search for them.  It would be nicer to be able to ask for a list of SDRAMPartition separately given that they are treated quite differently otherwise.

I agree that I haven't seen an example of bi-directional communication yet, so that isn't a requirement.

**Comment by rowleya** [comment]
I should add that I also think that that space having multiple sub-regions seems a bit specific a case to me.  I suspect the control of this will be a bit to specific to the use to make a general case here.  For example, what order should the writers to the space be in?  It isn't clear that this can be solved with general code that doesn't know how the application works in practice.

**Comment by alan-stokes** [comment]
so to treat 3 first (as that's the simplest) we could just add a new dict in the graph to keep them separate and thus you'd not need to search though all the mc ones too (id also be inclined to rename outgoing_partitions to mc ones for clarity)

on your use case. do the multiple srcs need the same sdram? or continuous sdram? or just some sdram? if its just some sdram, then its fine in my setup, you just have a separate outgoing partition for each src vertex to cover each edge. as it would be already. 

if you need the same sdram or a continuous block to be allocated in chunks by each src......... as it currently stands, i dont think it would support that. you'd need a sdram incoming partition.... which makes no sense...... that would need thinking over. I guess youd need to define it as a app sdram edge whos during partitioning (given the usecase youve got actually suits my partitioning framework better than it currently stands) partitions the overall sdram to each src. clunky...... but first pass over a idea

every use case i've faced have been single src and multiple targets. so i admit multiple src single target hasn't been considered.

**Comment by alan-stokes** [comment]
on "I should add that I also think that that space having multiple sub-regions seems a bit specific a case to me. I suspect the control of this will be a bit to specific to the use to make a general case here. For example, what order should the writers to the space be in? It isn't clear that this can be solved with general code that doesn't know how the application works in practice."

i admit this side hasnt been thought out much. it worked for attempting to understand nengo, and that was as far as i needed it to go. if we just want a accum like the ear uses. thats fine by me.

**Comment by rowleya** [comment]
> so to treat 3 first (as that's the simplest) we could just add a new dict in the graph to keep them separate and thus you'd not need to search though all the mc ones too (id also be inclined to rename outgoing_partitions to mc ones for clarity)
> 
> on your use case. do the multiple srcs need the same sdram? or continuous sdram? or just some sdram? if its just some sdram, then its fine in my setup, you just have a separate outgoing partition for each src vertex to cover each edge. as it would be already.
> 
> if you need the same sdram or a continuous block to be allocated in chunks by each src......... as it currently stands, i dont think it would support that. you'd need a sdram incoming partition.... which makes no sense...... that would need thinking over. I guess youd need to define it as a app sdram edge whos during partitioning (given the usecase youve got actually suits my partitioning framework better than it currently stands) partitions the overall sdram to each src. clunky...... but first pass over a idea
> 
> every use case i've faced have been single src and multiple targets. so i admit multiple src single target hasn't been considered.

I guess this would be multiple sources to multiple SDRAM when I think about it as otherwise you would get overwriting, so I guess this is really single source single target.  When I think about it, I can't see any case where you would have the same target for any more than one source, unless you are talking about a multi-region as you mentioned, where all the sources write to different parts of the same target (which would work for the neuron model, but I don't think it is needed).  I guess you get a sort-of multicast from having multiple targets read the same source though; in that case I guess the partition *is* valid...

**Comment by alan-stokes** [comment]
having eaten and drank. we could make the sdram partition exactly that, not incoming or outgoing, but a partition. which takes a list of src and dest and mods the src/dest's base address accoridngly. if you had both lists greater than 1 in size. i have no idea how that works.... but either being set to 1 would be doable

**Comment by alan-stokes** [comment]
so if you had say 3 synapse cores who write to a given continiuous sdram space which the dest reads as a dma and considers it some form of ring buffer.....but i feel im twisting a clinder to fit a twister here.

**Comment by alan-stokes** [comment]
points needed to fix.
1. placer resource tracker needs updating to handle edge cost
 2. auto pause and resume needs updating to handle edge cost
3. example needs augmenting to show multi sdram partitions

**Comment by alan-stokes** [comment]
4. augment graph to seperate outpug partitions and sdrma partitions

**Comment by Christian-B** [comment]
I would augment the graph to handle costed and non costed partitions.

The fact that currently the only costed ones are SDRam is  possibly remporary

**Comment by alan-stokes** [comment]
_outgoing_costed_edge_partitions_by_pre_vertex it be

**Comment by alan-stokes** [comment]
should i make a inhertance class for "costedPartition" which sdram inhirts from? t would allow the graph to ahndle whatever else it faces in the same vain

**Comment by Christian-B** [comment]
An Idea is not not make the SDRAMPartitions OutgoingEdgePartions at all.
Instead have both inherit from a shared Partition class which has most of the functionality except.

Source Vertex:  OutgoingEdgePartions has one, SDRAMPartitions can have more than one

Cost models: OutgoingEdgePartions have none, SDRAMPartitions do

PartitionAllocator: OutgoingEdgePartions have no such requirement, SDRAMPartitions do.

ect

**Comment by alan-stokes** [comment]
im not following,. and we said that unless we get confirmation that luca needs those sdram edges to be continious. theres no need to for multiple srcs. 

how about i finish what im doing. then you can fiddle in your own branch

**Comment by rowleya** [comment]
OK, I have been in contact with Luca.  His model is indeed that multiple sources write to a single contiguous space.  This is pretty much the reverse of the example where multiple targets read from a single space.  What I haven't confirmed is whether the single space is absolutely required or just convenient for doing a single read vs. multiple reads.

**Comment by alan-stokes** [comment]
well the new form of this now allows lucas model. as well as a multiple src multiple destination impl. Which only makes sense if say you had multiple neuron cores. but its there at least. ive still gotta fix the resource tracker bits. and auto pause and resume bits which were missed before. but ive added your ability to jsut search the "costed" partitions

File: pacman/model/graphs/abstract_multiple_partition.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+from pacman.exceptions import PacmanConfigurationException
+from spinn_utilities.ordered_default_dict import DefaultOrderedDict
+from spinn_utilities.ordered_set import OrderedSet
+
+
+class AbstractMultiplePartition(object):
+
+    def __init__(self, pre_vertices):
+        self._pre_vertices = OrderedDict()
```
Comment (Christian-B) [inline pacman/model/graphs/abstract_multiple_partition.py]: I would use either pre and post OR source and destination. The mix is confusing.

File: pacman/model/graphs/abstract_costed_partition.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from six import add_metaclass
+from spinn_utilities.abstract_base import AbstractBase
+
+
+@add_metaclass(AbstractBase)
+class AbstractCostedPartition(object):
+
```
Comment (Christian-B) [inline pacman/model/graphs/abstract_costed_partition.py]: Slots?

Comment (Christian-B) [reply pacman/model/graphs/abstract_costed_partition.py]: total_sdram_requirements function???

Comment (alan-stokes) [reply pacman/model/graphs/abstract_costed_partition.py]: doing that causes a dimond inhertance. i think. will need to look more closly after merge

File: pacman/model/graphs/abstract_sdram_partition.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from six import add_metaclass
+from spinn_utilities.abstract_base import abstractmethod, AbstractBase
+
+
+@add_metaclass(AbstractBase)
+class AbstractSDRAMPartition(object):
```
[the remark this answers is not in the document]
Comment (alan-stokes) [reply pacman/model/graphs/abstract_sdram_partition.py]: not true. esp as SegmentedSDRAMMachinePartition is a final class in all this. 

there is constant sdram, dest seg sdram, src seg sdram, machine, application (outgoing partitions) which are the final objects.

the only things that inheirt from sdram is the 3 partitions that do sdram, and they have to inheit from signle and multile accordingly to their behaviour. So this question makes no sense

File: pacman/model/graphs/impl/constant_sdram_machine_partition.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        AbstractTrafficTypeSecureOutgoingPartition.__init__(
+            self,  identifier=identifier,
+            allowed_edge_types=SDRAMMachineEdge,
+            label=label, traffic_type=EdgeTrafficType.SDRAM)
+        AbstractSDRAMPartition.__init__(self)
+        self._sdram_base_address = None
+        self._pre_vertex = pre_vertex
+
+    @property
+    def pre_vertex(self):
```
[the remark this answers is not in the document]
Comment (alan-stokes) [reply pacman/model/graphs/impl/constant_sdram_machine_partition.py]: no code is shareable. the consant has to ensure all edges have the same cost, and the base address is the same for each edge. whereas the the segged have to accum the sdram edge cost and then have to allocate base addresses differnetly if its a src or dest seg. there is no cloned code in this view.

[the remark this answers is not in the document]
Comment (alan-stokes) [reply pacman/model/graphs/impl/constant_sdram_machine_partition.py]: esp on pre vertex. which is forced by the abstract single src partition. so its already grouped if your on about that speicifc bit

File: pacman/operations/placer_algorithms/spreader_placer.py [hunk 5]
```diff
@@ -13,10 +13,10 @@
 # You should have received a copy of the GNU General Public License
 # along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
-from collections import defaultdict
-import functools
-import math
-import sys
+try:
+    from collections.abc import defaultdict
```
[the remark this answers is not in the document]
Comment (alan-stokes) [reply pacman/operations/placer_algorithms/spreader_placer.py]: technically this has nothing to do with this pr. but will fix for you

File: pacman/model/routing_info/dict_based_machine_partition_n_keys_map.py [hunk 6]
```diff
@@ -45,4 +45,5 @@ def set_n_keys_for_partition(self, partition, n_keys):
 
     @overrides(AbstractMachinePartitionNKeysMap.n_keys_for_partition)
     def n_keys_for_partition(self, partition):
-        return self._n_keys_map[partition]
+        if partition in self._n_keys_map:
```
[the remark this answers is not in the document]
Comment (alan-stokes) [reply pacman/model/routing_info/dict_based_machine_partition_n_keys_map.py]: think the better apporach is to actually raise an exception. as if your asking for n keys and its not been allocated keys, there's already more issues going on. Im suspicious that this change was a debug at some point .

[the remark this answers is not in the document]
Comment (alan-stokes) [reply pacman/model/routing_info/dict_based_machine_partition_n_keys_map.py]: im assuming its becfore i fixed all the pacman calls to check if the edge type was multicast

[the remark this answers is not in the document]
Comment (alan-stokes) [reply pacman/model/routing_info/dict_based_machine_partition_n_keys_map.py]: fixed in latest commit

File: pacman/operations/placer_algorithms/spreader_placer.py [hunk 7]
```diff
@@ -26,6 +26,9 @@
 from pacman.utilities.utility_objs import ResourceTracker
 from pacman.model.constraints.placer_constraints import (
     SameChipAsConstraint, ChipAndCoreConstraint)
+import functools
```
[the remark this answers is not in the document]
Comment (alan-stokes) [reply pacman/operations/placer_algorithms/spreader_placer.py]: moved. im going to assume pycharm doesnt follow whatever hoop it is for top vs bottom

File: pacman/model/graphs/abstract_edge_partition.py [hunk 8]
```diff
@@ -19,7 +19,7 @@
 
 
 @add_metaclass(AbstractBase)
-class AbstractOutgoingEdgePartition(object):
+class AbstractEdgePartition(object):
```
[the remark this answers is not in the document]
Comment (alan-stokes) [reply pacman/model/graphs/abstract_edge_partition.py]: was a orginal api force *points at rowley who asked for these long long time ago* removed as discussed

File: pacman/model/graphs/abstract_multiple_partition.py (hunk 1 above)
Comment (alan-stokes) [reply pacman/model/graphs/abstract_multiple_partition.py]: its prevertices as the only thing that handles multiple pre is the src segged which this api allows to work in a general case. 

pre vertices is what the graph handles, so keeping in the same vain makes it sensible there. multiple dests are handled by the edges of a partition.

File: pacman/model/graphs/abstract_costed_partition.py (hunk 2 above)
Comment (alan-stokes) [reply pacman/model/graphs/abstract_costed_partition.py]: so the issue here is that i dont know what other costed stuff we'll ever have. thats why ive left it as a separate thing

File: pacman/model/graphs/abstract_application_outgoing_partition.py [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+from six import add_metaclass
+
+from spinn_utilities.abstract_base import abstractmethod, AbstractBase
+
+
+@add_metaclass(AbstractBase)
+class AbstractApplicationOutgoingPartition(object):
```
[the remark this answers is not in the document]
Comment (alan-stokes) [reply pacman/model/graphs/abstract_application_outgoing_partition.py]: this adds a api which is only acceptable in the app levle. not amchine level. so needs to not go into outgoing edge partition

File: pacman/model/graphs/abstract_single_source_partition.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from pacman.exceptions import PacmanConfigurationException
+
+
+class AbstractSingleSourcePartition(object):
```
[the remark this answers is not in the document]
Comment (alan-stokes) [reply pacman/model/graphs/abstract_single_source_partition.py]: fixed

**Comment by dkfellows** [comment]
I'm generally in favour of SDRAM edges, _especially_ as machine edges.

File: pacman/utilities/json_utils.py [hunk 11]
```diff
@@ -330,6 +331,10 @@ def graph_from_json(json_dict):
         vertex_add_contstraints_from_json(j_vertex, graph)
     for j_edge in json_dict["edges"]:
         edge = edge_from_json(j_edge, graph)
+        if not graph.outgoing_partition_exists(edge.pre_vertex, "JSON_MOCK"):
```
Comment (dkfellows) [inline pacman/utilities/json_utils.py]: WTF is test code doing in a non-test class?

File: pacman/model/graphs/abstract_application_outgoing_partition.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Christian-B) [reply pacman/model/graphs/abstract_application_outgoing_partition.py]: I understand the desire to have ApplicationOutgoingPartition and MachineOutgoingEdgePartition.

 However adding and Abstract fir each of these where the above is the only implementation and this is HARD coded into MachineGrap and ApplicationGraph is just Abstract class for the sake of having useless extra BullShit!

File: pacman/utilities/json_utils.py (hunk 11 above)
Comment (alan-stokes) [reply pacman/utilities/json_utils.py]: i have absolutely no idea.........

**Comment by dkfellows** [comment]
I _think_ I've managed to merge this correctly past the graph mapper changes. However, there were multiple significant conflicting merges done to get to this point (which I've resolved and partially squashed) so this ***needs*** to be re-reviewed by @alan-stokes and @Christian-B to check if it is even basically correct before considering whether it is a change we wish to have.
